Add industrial profile

rojkov opened the pull request.

rojkov commented:
Oops... Turned out this PR depends on https://github.com/bmwcarit/meta-ros/pull/457 and one more new commit in meta-openembedded (http://cgit.openembedded.org/meta-openembedded/commit/?id=e8d5a1bb6a022172246f095aca0979ab5a6bd2ed).

mythi commented:
@rojkov looks like svn is needed in ```HOSTTOOLS``` to get builds ongoing.

rojkov commented:
test this please

rojkov commented:
test this please

rojkov commented:
test this please
(dropped python-empy)

rojkov commented:
test this please

okartau commented:
test this please. I dont know why some requests fail to trigger build

rojkov commented:
test this please

rojkov commented:
* rebased to the latest https://github.com/intel/intel-iot-refkit/pull/109 and meta-ros;
* force pushed.

rojkov commented:
* made this PR independent from #109 again;
* updated meta-ros;
* added demos to meta-refkit-extra;
* more detailed `doc/industrial-profile.rst`
* rebased and force pushed.

gvancuts commented on doc/industrial-profile.rst:
```diff
@@ -0,0 +1,85 @@
+#. Pick and place.
+
+However combined with the computer vision profile they can be extended to
+other applications like drons, self-driving cars, agricultural robots etc.
```
s/drons/drones

gvancuts commented on doc/industrial-profile.rst:
```diff
@@ -0,0 +1,85 @@
+  joints movement planning,
+- core ROS-I components used for interfacing with real robot hardware.
+
+Due to the embedded nature of Refkit no GUI components have been included
```
Should we use 'IoT RefKit' instead of 'Refkit'?

gvancuts commented on doc/industrial-profile.rst:
```diff
@@ -0,0 +1,85 @@
+
+For detailed usage instructions please refer `ROS Industrial tutorials`_
+
+How to use
```
Does this section apply to the platform running the IoT RefKit?

gvancuts commented on meta-refkit-extra/doc/robotics-uarm-metal.rst:
```diff
@@ -0,0 +1,284 @@
+
+G205 <base_angle> <left_angle> <right_angle> <hand_angle>
+   Set position for all four servos. The code accepts four parameters which
+   are PWM signal widths for the respective servos in miliseconds.
```
s/miliseconds/milliseconds

gvancuts commented on meta-refkit-extra/doc/robotics-uarm-metal.rst:
```diff
@@ -0,0 +1,284 @@
+Setting up
+==========
+
+1. Make sure you have installed the packages `uarmmeta--moveit-config`,
```
How can one check if they are installed and install them if they are not?

rojkov commented:
* addressed @gvancuts 's review comments;
* rebased.

gvancuts commented on meta-refkit-extra/doc/robotics-uarm-metal.rst:
```diff
@@ -0,0 +1,293 @@
+Now your mechanical arm can be controlled through MoveIt's Move Group Interface.
+It's possible to control the arm interactively  with a GUI ROS node installed
+on your desktop. For that copy the source code of the packages `uarmmetal-support`
+and `uarmmetal-moveit-configs` to your catkin workspace on the desktop as
```
I know I may sound ignorant but I had no idea what a catkin workspace was... How about adding a URL (e.g http://wiki.ros.org/catkin) in the text that would provide a pointer to the other less-informed out there ;)

rojkov replied:
Ignorance is bliss :) I got blind to issues of this kind.

gvancuts commented on meta-refkit-extra/doc/robotics-uarm-metal.rst:
```diff
@@ -0,0 +1,293 @@
+
+     $ bitbake refkit-image-common
+
+2. Plug the mechanical arm in a USB port.
```
I know it may sound obvious but can you add one step to indicate that the image should be installed on the target platform (and if we have a standard installation guide, you could point to it).

gvancuts commented on meta-refkit-industrial/recipes-image/images/refkit-image-industrial.bb:
```diff
@@ -0,0 +1,17 @@
+SUMMARY = "PROJECTNAME image for Industrial Robotics profile."
```
Out of curiosity, is 'PROJECTNAME' supposed to be expanded/replaced at some point?

rojkov commented:
* addressed review comments;
* rebased.

gvancuts commented on meta-refkit-industrial/recipes-industrial/packagegroups/packagegroup-industrial.bb:
```diff
@@ -0,0 +1,14 @@
+SUMMARY = "PROJECTNAME Industrial package groups"
```
Should you also swap PROJECTNAME for IoT Reference OS Kit (as you did in the refkit-image-industrial.bb file)?

gvancuts commented on meta-refkit-industrial/recipes-industrial/packagegroups/packagegroup-industrial.bb:
```diff
@@ -0,0 +1,14 @@
+
+inherit packagegroup
+
+SUMMARY_${PN} = "PROJECTNAME Industrial profile"
```
Should you also swap PROJECTNAME for IoT Reference OS Kit (as you did in the refkit-image-industrial.bb file)?

rojkov commented:
Replaced missed PROJECTNAME placeholders with "IoT Reference OS Kit".

mythi commented on meta-refkit-industrial/recipes-ros/cv-bridge/cv-bridge_%.bbappend:
```diff
@@ -0,0 +1,3 @@
+# TODO: This way cv_bridge doesn't require libboost_python, but this tweak
+#       should be dropped after ROS is compatible with python3-only distro.
+EXTRA_OECMAKE_append_refkit-industrial = " -DANDROID=ON"
```
Does this mean #109 isn't absolutely necessary for this anymore? What are the limitations after this change?

rojkov replied:
#109 is unrelated to this. Without this option cv_bridge wants to build python2 bindings which depend on python2 bindings to Boost (disabled in oe-core). The proper solution would be to rewrite cv_bridge's code to support python3.

mythi commented:
@rojkov is the image intentionally not built?

rojkov commented:
Not intentionally. For some reason I thought the image is included in build targets automatically. Need to check what's needed to build it in CI.

mythi commented on meta-refkit/conf/distro/include/refkit-ci.inc:
```diff
@@ -70,7 +70,7 @@ REFKIT_CI_SDK_TARGETS=""
 REFKIT_CI_ESDK_TARGETS=""
 # Following targets would be executed with do_test_iot_export task
-REFKIT_CI_TEST_EXPORT_TARGETS="refkit-image-common refkit-image-computervision refkit-image-gateway"
+REFKIT_CI_TEST_EXPORT_TARGETS="refkit-image-common refkit-image-computervision refkit-image-gateway refkit-image-industrial"
```
My preference is to make it explicit in ```REFKIT_CI_BUILD_TARGETS```.  The same for refkit-image-computervision and refkit-image-gateway now that the the comment lines (which could be removed) before  ```REFKIT_CI_BUILD_TARGETS``` is obsolete.

Additionally, I think adding refkit-image-industrial here is not needed until we're ready to test it.

rojkov commented:
Ok. Added the profile images to `REFKIT_CI_BUILD_TARGETS` explicitly.

rojkov commented:
test this please

rojkov commented:
test this please

ipuustin commented on doc/profiles.rst:
```diff
@@ -61,3 +61,4 @@ Profile Summary
 #. refkit-image-common: A configuration with interactive tools but without any special software in it. Recommended target if you are not interested in any given profile.
 #. refkit-image-computervision: A profile with tools and configuration for computer vision use cases.
 #. refkit-image-gateway: A profile with tools and configuration for acting as an IoT sensor.
+#. refkit-image-industrial: A profile with core ROS Industrial packages intended to be used as a base for industrial robotics use cases.
```
This is just a thought, but should the image (and profile) rather be called `industrial-robotics`?

ipuustin commented on meta-refkit/conf/distro/include/refkit-ci.inc:
```diff
@@ -59,9 +59,15 @@ REFKIT_CI_POSTBUILD_SELFTESTS="secureboot refkit-poky refkit-license-check image
+REFKIT_CI_BUILD_TARGETS=" \
+  refkit-image-minimal \
+  refkit-image-computervision \
+  refkit-image-gateway \
```
Are `refkit-image-computervision` and `refkit-image-gateway` added on purpose?

ipuustin commented:
There should be an explanation in the commit message why the change from TinyXML to TinyXML2 is needed, and some commentary about when the patch set can be merged into upstream layers.

ipuustin commented on meta-refkit/recipes-ros/cmusphinx/sphinxbase_0.8.bbappend:
```diff
@@ -0,0 +1,3 @@
+DEPENDS = "libsndfile1 alsa-lib libsamplerate0"
```
Shouldn't this bbappend be in `meta-refkit-industrial`? This might be worth doing an upstream patch for a `PACKAGECONFIG[python]` or similar.

ipuustin commented:
Well, it appears that many files are in `meta-refkit/recipes-ros`. The whole `recipes-ros` directory might be moved to `meta-refkit-industrial`.

ipuustin commented on meta-refkit/recipes-ros/cmake-modules/cmake-modules/0001-Add-FindTinyXML2-module.patch:
```diff
@@ -0,0 +1,96 @@
+Date: Fri, 3 Jun 2016 14:29:45 +0300
+Subject: [PATCH] Add FindTinyXML2 module
+
+Upstream-Status: Submitted [https://github.com/ros/cmake_modules/pull/42]
```
Merged already...

ipuustin commented on meta-refkit/recipes-ros/control-toolbox/control-toolbox/0001-Switch-from-TinyXML-to-TinyXML2.patch:
```diff
@@ -0,0 +1,173 @@
+Date: Thu, 26 Jan 2017 14:42:31 +0200
+Subject: [PATCH] Switch from TinyXML to TinyXML2
+
+Upstream-Status: Inappropriate [ROS won't move to TinyXML2 until kinetic]
```
What is kinetic?

ipuustin commented on meta-refkit/recipes-ros/moveit/moveit-ros-planning/0002-moveit_ros-planning-Use-TinyXML2-instead-of-TinyXML.patch:
```diff
@@ -0,0 +1,83 @@
+Date: Wed, 25 Jan 2017 11:06:58 +0200
+Subject: [PATCH 2/2] moveit_ros/planning: Use TinyXML2 instead of TinyXML
+
+Upstream-Status: Inappropriate [TinyXML2 is WONTFIX in indigo-devel]
```
Add a link here to the issue. At least I'm interested if we can ever upstream this... I don't even know what is `indigo-devel`.

ipuustin commented on meta-refkit/recipes-ros/ros-control/controller-manager/0001-controller_manager-drop-unused-inclusion-of-tinyxml..patch:
```diff
@@ -0,0 +1,27 @@
+Date: Thu, 26 Jan 2017 14:30:04 +0200
+Subject: [PATCH] controller_manager: drop unused inclusion of tinyxml.h
+
+Upstream-Status: Submitted [https://github.com/ros-controls/ros_control/pull/261]
```
Merged also.

ipuustin commented on meta-refkit/conf/distro/include/refkit-supported-recipes.txt:
```diff
@@ -161,6 +161,11 @@ i2c-quark-board@quark-bsp
 iftop@networking-layer
 image-transport@ros-layer
+industrial-msgs@ros-layer
+industrial-robot-client@ros-layer
```
It might make sense to merge this patch with the previous one that was changing the same file.

ipuustin commented on meta-refkit/recipes-core/packagegroups/packagegroup-industrial.bb:
```diff
@@ -0,0 +1,14 @@
+inherit packagegroup
+
+SUMMARY_${PN} = "IoT Reference OS Kit Industrial Robotics profile"
+RDEPENDS_${PN} = "\
```
Do you need a separate packagegroup for development images? Or are the packages the same in both?

ipuustin commented on meta-refkit/recipes-ros/cv-bridge/cv-bridge_%.bbappend:
```diff
@@ -0,0 +1,3 @@
+# TODO: This way cv_bridge doesn't require libboost_python, but this tweak
+#       should be dropped after ROS is compatible with python3-only distro.
+EXTRA_OECMAKE += "-DANDROID=ON"
```
Hmm, the next patch says: `switch ROS to python3 `. Is it python3-compatible or not?

ipuustin commented:
I think the PR should be amended with the changes in patch `meta-refkit-industrial: make the layer Yocto 2.0 compatible`, and then the patch should be dropped. The patches shouldn't fix previous patches in the same PR.

mythi commented:
On Tue, Jun 13, 2017 at 5:12 PM, Ismo Puustinen wrote:

> I think the PR should be amended with the changes in patch meta-refkit-industrial:
> make the layer Yocto 2.0 compatible, and then the patch should be
> dropped. The patches shouldn't fix previous patches in the same PR.
>
>
+1

btw, it'd be easy to add meta-refkit-industrial for compatibility checking
too (see the automatically generated tests in refkit-poky.py)

rojkov commented:
Initially I thought that the layer might have packages unrelated to
robotics, e.g. LinuxCNC or MachineKit. I wouldn't like to narrow the scope
still.

On Tue, Jun 13, 2017 at 2:58 PM, Ismo Puustinen <notifications@github.com>
wrote:

> *@ipuustin* commented on this pull request.
> ------------------------------
>
> In doc/profiles.rst
> <https://github.com/intel/intel-iot-refkit/pull/49#discussion_r121655328>:
>
> > @@ -61,3 +61,4 @@ Profile Summary
>  #. refkit-image-common: A configuration with interactive tools but without any special software in it. Recommended target if you are not interested in any given profile.
>  #. refkit-image-computervision: A profile with tools and configuration for computer vision use cases.
>  #. refkit-image-gateway: A profile with tools and configuration for acting as an IoT sensor.
> +#. refkit-image-industrial: A profile with core ROS Industrial packages intended to be used as a base for industrial robotics use cases.
>
> This is just a thought, but should the image (and profile) rather be
> called industrial-robotics?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/intel/intel-iot-refkit/pull/49#pullrequestreview-43699450>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACCYQH2GFUXRLG6PAYoM6Cw_VIKOsY6ks5sDnlSgaJpZM4MM2Pt>
> .
>

rojkov commented:
Yes. @mynti asked to include them explicitly to built targets in the scope
of this PR.

On Tue, Jun 13, 2017 at 3:03 PM, Ismo Puustinen <notifications@github.com>
wrote:

> *@ipuustin* commented on this pull request.
> ------------------------------
>
> In meta-refkit/conf/distro/include/refkit-ci.inc
> <https://github.com/intel/intel-iot-refkit/pull/49#discussion_r121656316>:
>
> > @@ -59,9 +59,15 @@ REFKIT_CI_POSTBUILD_SELFTESTS="secureboot refkit-poky refkit-license-check image
>  # Any other symbols would be skipped in parser.
>  #
>  # Following targets would be used to perform default build task.
> -# When adding new profile images, add them to refkit-installer-image.bb
> -# and they will get pulled into the build indirectly.
> -REFKIT_CI_BUILD_TARGETS="refkit-image-minimal refkit-installer-image swtpm-wrappers-native ovmf"
> +REFKIT_CI_BUILD_TARGETS=" \
> +  refkit-image-minimal \
> +  refkit-image-computervision \
> +  refkit-image-gateway \
>
> Are refkit-image-computervision and refkit-image-gateway added on purpose?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/intel/intel-iot-refkit/pull/49#pullrequestreview-43700546>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACCYamK9otIv7YJkZDW616dPg0NFR4tks5sDnqvgaJpZM4MM2Pt>
> .
>

rojkov commented:
I must confess I don't fully understand what you imply here. I just tried
to follow the way how packages are included in other profiles' images.
Mindlessly though.

On Tue, Jun 13, 2017 at 5:08 PM, Ismo Puustinen <notifications@github.com>
wrote:

> *@ipuustin* commented on this pull request.
> ------------------------------
>
> In meta-refkit/recipes-core/packagegroups/packagegroup-industrial.bb
> <https://github.com/intel/intel-iot-refkit/pull/49#discussion_r121688237>:
>
> > @@ -0,0 +1,14 @@
> +SUMMARY = "IoT Reference OS Kit Industrial Robotics profile package groups"
> +LICENSE = "MIT"
> +PR = "r1"
> +
> +inherit packagegroup
> +
> +SUMMARY_${PN} = "IoT Reference OS Kit Industrial Robotics profile"
> +RDEPENDS_${PN} = "\
>
> Do you need a separate packagegroup for development images? Or are the
> packages the same in both?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/intel/intel-iot-refkit/pull/49#pullrequestreview-43734489>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACCYRcFnciUxYJj8UwZDy2LMm26v1XYks5sDpfCgaJpZM4MM2Pt>
> .
>

rojkov commented:
Thanks @ipuustin for the review!

* split commits adding profile images to build targets;
* reorganized the PR not to fix commits in the same PR (the new layer is Yocto 2.0 compatible from the beginning now);
* updated and added missing references to upstream submissions for the "Switch from TinyXML to TinyXML2" patch set;
* dropped some patches as the packages they change have been excluded from the profile and are not  used in the demos;
* clarified the comment in `cv-bridge_%.bbappend` a bit about python2/3 compatibility in future ROS;
* force pushed.

Arthil commented:
There was a problem in CI, where one of the boards froze at test phase. Should be fixed now. 
Test this please.

mythi commented:
@rojkov looks nice! please rebase and I'll create a combined PR
